### ISSUE TITLE
Localisation: Russian

### ISSUE THREAD

**Issue by JuliusSweetland** [OptiKey/OptiKey#150]
**1.Transifex translation of UI strings:** Complete

**2.Keyboard layout:**
"Russian, Belarusian and Ukranian standard keyboards"...
![image](https://cloud.githubusercontent.com/assets/4722561/10587646/f41d7b30-7699-11e5-970c-865ed5b6d76d.png)

"Russian (Typewriter) keyboard"...
![image](https://cloud.githubusercontent.com/assets/4722561/10587659/0493cdc0-769a-11e5-8af9-f8faa78f1b27.png)

Special characters that need to be easy to get to: *can I have some input on this?*

**3.Word list:** included on the branch

**4.Anything else:** nothing at the moment

**BRANCH:** https://github.com/JuliusSweetland/OptiKey/tree/Russian (work done by @Ar2rZ)

**Comment by Ar2rZ** [comment]
![belarusian](https://cloud.githubusercontent.com/assets/14283948/10602807/72a75598-7712-11e5-9d75-16ae2ae5cd04.png)
![ukranian](https://cloud.githubusercontent.com/assets/14283948/10602814/77d4fd0e-7712-11e5-8907-b9f443969151.png)

**Comment by Ar2rZ** [comment]
3.Word list: included on the branch
will try to find dictionaries for Ukrainian and for Belorussian languages.

1.Transifex translation of UI strings: 96% - a few strings remain
left strings are not required to translate.. as per my understanding...or?

**Comment by Ar2rZ** [comment]
functionality for changing layouts on the go? in this case we can leave for moment Ukrainian and Belorussian translation till some body will be able to translate, can be used Russian, as in those countries everybody is speaking Russian.

**Comment by JuliusSweetland** [comment]
@Ar2rZ I've updated the UI string translation status to 'complete' as you're correct. The remaining strings should be left out.

At the moment I haven't thought of a way to change layouts on the fly (as you suggested that we also need a way to switch between Russian and English, for example). Let's leave that problem for the moment and we can return to it when we have a solution.

**Comment by Ar2rZ** [comment]
translation for Russian will require to improve.. 
as i had no time to test this weekend, will try to this this weekend.
how we will deal with those changes made for other functionalities included into other languages layouts? as in Russian it will quiet difficult to add more columns..

**Comment by JuliusSweetland** [comment]
@Ar2rZ No rush. I'll aim to merge in master at some point this week which will add 1 new column. Hopefully that doesn't make the keys too small, but we'll have to see. I'll post images on here and we can decide.

**Comment by Ar2rZ** [comment]
what i am worrying about - how it will looks like and how  good eyetracker functionality is on smaller screens?
 as i am testing on 24" display....

**Comment by JuliusSweetland** [comment]
I'll do my best and then we can review together. I test on a 15.6" screen
so if I can make it work we should be ok

**Comment by Ar2rZ** [comment]
ok, thanks.

so, Ukrainian and Belorussian. what we will do about them? 
once you finish merge with latest changes, i can include those layouts as well, but ether leave them disabled, or for general translation using Russian, as they will be quiet close, till we have those languages translated..

**Comment by JuliusSweetland** [comment]
Let's leave them out and only include Russian for now. We can complete
Russian and then branch again to include these as there will be
accompanying changes.

**Comment by JuliusSweetland** [comment]
@Ar2rZ Are there any special characters in Russian beyond those on the ALPHA keyboard? I mean characters that should be first on the DIACRITICS keyboard so that they are easy to access.

**Comment by Ar2rZ** [comment]
no. all letters are on main keyboard.

**Comment by JuliusSweetland** [comment]
I've just submitted changes to the Russian branch to add keyboard keys to the ALPHA keyboard. The only obvious issue is that the text is now very small on all keys (possibly from how long the Russian string for "Multi-key selection" is. Also the text size between the top keys (e.g. CLEAR) and the bottom keys is out of alignment, but this might be down to something else - I'll investigate tomorrow.

**Comment by Ar2rZ** [comment]
i did some changes in translation for alpha keyboard.
now they are aligned correctly. but font still smaller then in English..  i don't see the reason why is should be smaller, 
one think more - i can't shorten other words, as it  will loose context meaning.

**Comment by Ar2rZ** [comment]
![001](https://cloud.githubusercontent.com/assets/14283948/10662454/c16cdee6-78ac-11e5-88d9-290748c46acd.png)

**Comment by Ar2rZ** [comment]
question:
 why on Alpha keyboard is button leading to Alpha?

**Comment by JuliusSweetland** [comment]
> why on Alpha keyboard is button leading to Alpha?

@Ar2rZ It's a temporary issue. I have a new keyboard which I would like to create and add to all other keyboards, which will replace the key pointing to the current keyboard. Until I've created the new keyboard I've just included a self-reference key to keep the layout balanced. It will disappear soon.

**Comment by JuliusSweetland** [comment]
>now they are aligned correctly. but font still smaller then in English.. i don't see the reason why is should be smaller

@Ar2rZ Great thanks. I'll check it out and see if I can figure out what's causing the small text. The sizing of the text is a complicated beast - it relies on the SharedSizeGroup dependency property which attempts to keep grid column sizes in line with each other, as well as the zooming control called a ViewBox. Together they produce difficult to understand behaviour, but I'll dig into it and see what I can find.

**Comment by Ar2rZ** [comment]
another thing - capturing filling pie is not in the middle of the key just a bit left. is seems to be on other languages is same.
![002](https://cloud.githubusercontent.com/assets/14283948/10663505/efd7fce6-78b3-11e5-94c5-6b12ea36c9f9.png)
![003](https://cloud.githubusercontent.com/assets/14283948/10663508/f395bd14-78b3-11e5-8a02-3c995e53b3db.png)

**Comment by JuliusSweetland** [comment]
Good spot - I've just fixed this on master of you want to update from
master?

**Comment by Ar2rZ** [comment]
does not matter at the moment..

**Comment by JuliusSweetland** [comment]
Ok, but in the GitHub application you can update from master at any point
with one click and everything should sync automatically at the moment.

**Comment by Ar2rZ** [comment]
ok. thanks.

**Comment by Ar2rZ** [comment]
i think, filling pie on arrow has to be on the arrow pointing not on the middle of the arrow.
is very distractive, from my point of view..

**Comment by JuliusSweetland** [comment]
I agree, but it's not really possible as the corner of the pop-up has to be
at the point of interest. Imagine the point of interest is in the very
corner of the screen - the pie animation would be 3/4 cut off by the edge
of the screen.

**Comment by Ar2rZ** [comment]
maybe is possible then make the pie more transparent?

**Comment by JuliusSweetland** [comment]
I'll experiment and see how it looks

**Comment by Ar2rZ** [comment]
may be you can point me where i can change the transparency of the pie animation?

**Comment by JuliusSweetland** [comment]
In one of the theme resource files, (e.g. Android_Dark.xaml) you can edit the PointSelectionProgressStyle to set the opacity of the progress pie animation. This can be edited as below (the extra bold line):

````
            <Style x:Key="PointSelectionProgressStyle" TargetType="{x:Type controls:ProgressIndicator}">
                **<Setter Property="Opacity" Value="0.5" />**
                <Setter Property="Fill" Value="{StaticResource PointSelectionProgressBrush}" />
                <Setter Property="MaxValue" Value="1" />
                <Setter Property="Value" Value="{Binding Path=SelectionProgress, RelativeSource={RelativeSource AncestorType={x:Type controls:CursorPopup}}, Mode=OneWay}" />
                <Style.Triggers>
                    <!--Progress is 0-->
                    <Trigger Property="Value" Value="0">
                        <Setter Property="Visibility" Value="Hidden" />
                    </Trigger>
                </Style.Triggers>
            </Style>
````

and looks like this:

![image](https://cloud.githubusercontent.com/assets/4722561/10712118/5056593e-7a88-11e5-8c4b-6b1078290080.png)

**Comment by Ar2rZ** [comment]
thanks.
will try.

**Comment by Ar2rZ** [comment]
may be is worth to put information with OptiKey support email and website links and other relevant info on the last tab on 'Settings'?

**Comment by JuliusSweetland** [comment]
Possibly. The website is on the splash screen.

**Comment by Ar2rZ** [comment]
Russian is ready. will create pull request for rexs files update later today.

**Comment by Ar2rZ** [comment]
@JuliusSweetland  do you have any other issues/suggestions?

**Comment by JuliusSweetland** [comment]
> Russian is ready. will create pull request for rexs files update later today.

@Ar2rZ Thanks Arthur. I'll take a look this evening. Is there anyone who can act as a reviewer if I create a pre-release candidate?

The only other thing I can think of is that for Russian no "letter" keys are simulated as real key presses, so it is not currently possible to perform a CTRL+P type of key press. If you can give me the key mappings (from Russian character to English character) that makes sense then we could try this?

**Comment by Ar2rZ** [comment]
![russiankeyboard](https://cloud.githubusercontent.com/assets/14283948/10756103/ded6636c-7c99-11e5-9498-b8ab71ecc0bb.png)

**Comment by Ar2rZ** [comment]
that what you was asking for?

**Comment by Ar2rZ** [comment]
yop, there will some body, who will review.

**Comment by JuliusSweetland** [comment]
Can you try an experiment and add mappings from the Russian letters to the
existing key code mapping method in the CharExtensions class and see how it
behaves when you perform ctrl+ key shortcuts? (Don't forget to release the
shift key when testing as it will probably be on by default.

**Comment by Ar2rZ** [comment]
ok, i will try now.

**Comment by Ar2rZ** [comment]
looks good. short keys are working as expected..

**Comment by JuliusSweetland** [comment]
Great. If you add to the pull request I'll take a look this evening.

**Comment by JuliusSweetland** [comment]
**Russian** Pre-release #1 available here: https://github.com/JuliusSweetland/OptiKey/releases/tag/v2.2.2_Russian_pre-release_1

**Comment by Ar2rZ** [comment]
thanks, will be reviewed in two-three days.

**Comment by JuliusSweetland** [comment]
No rush.

I thought of a potential issue. If OptiKey is simulating physical key
presses (e.g. pressing the "S" key when the "ы" char is selected), but the
OS language is not Russian, then OptiKey will show "ы" in the scratchpad,
but type "S" into the other application. I might add another setting to be
able to enable/disable physical key simulation. If it is disabled the
actual string (e.g. "ы") will be output instead. I think this is only
useful if the user's OS language differs from the language specified in
OptiKey (and the two languages are based on different character sets).

**Comment by Ar2rZ** [comment]
i see, i tested with mulitkey selection enabled. so, without multikey  is typing incorrect character..

**Comment by Ar2rZ** [comment]
it will useful  only for non Latin char sets..

**Comment by JuliusSweetland** [comment]
That's right, yes. I'll think more about the setting, but it might be
useful and it's an ready thing to add.

**Comment by Ar2rZ** [comment]
so, what is the plan? or this fix is part some thing else?

**Comment by JuliusSweetland** [comment]
Continue without it. I'll add it later.

**Comment by AnyKey-** [comment]
it requires a language switch layout. otherwise little use. To change the input language you need to go into the settings.

**Comment by JuliusSweetland** [comment]
@AnyKey- Do you mean OptiKey needs the ability to easily switch languages?

**Comment by AnyKey-** [comment]
yes of course

**Comment by JuliusSweetland** [comment]
What's the use case? To enter individual words in English while generally
typing in Russian (for example)?

**Comment by AnyKey-** [comment]
or make folder with english name and file in russian. or for people with disabilities. control only through the eyes of how to enter an address and Google search Russian? let's say if a person has no hands, how he pressed ALT+M?

**Comment by JuliusSweetland** [comment]
New issue with some initial thoughts here: https://github.com/JuliusSweetland/OptiKey/issues/167

I'm thinking of adding a new "LATIN"/"ENGLISH" keyboard which can be used to write english words. This keyboard would not support multi-key selection, or auto-complete, only one letter at a time typing (as this is the easiest thing to implement) and would only be available when the OptiKey language is not based on the latin alphabet, e.g. Russian. Would this meet your immediate needs?

**Comment by AnyKey-** [comment]
I propose to improve the product based on the vision of optimal usage. no urgent need. Only letter in English and Russian with auto-complete will be great.

**Comment by JuliusSweetland** [comment]
Switching languages presents a few challenges. A method of switching language would be required (which is fine), but it introduces this issue: https://github.com/JuliusSweetland/OptiKey/issues/168

I'll have to think about this for the moment.

**Comment by Ar2rZ** [comment]
so, if output of virtual keys are related to OS language, then win+space shortcut  will change output language, we can add English letter above Russian for example and will have proper English/Russian  output for the moment. and then later on can be add feature with multiple layouts...
i tested this on win 8.1. is working perfectly..

**Comment by Ar2rZ** [comment]
English  letter "D" will be in some other place.
 other letters seems to be in same place as in ordinary English layout.

**Comment by JuliusSweetland** [comment]
@Ar2rZ That's effective, but quite a complex approach. I need to do some experimenting so bear with me for a few days and I'll see what I can come up with. Ideally I'd like to add a new keyboard to the MENU called LANGUAGE which will allow you to switch the keyboard layout and output text, so a user could start OptiKey in Russian, switch to English UK to enter some english text, and then switch back, or even to another language if they want. I'd like the user to not have to be aware of the OS keyboard layout/language at all, as this introduces a really complicated concept. If I can hide all that then that would be ideal.

I'll update as soon as I have something.

Also regarding the "D" key: lots of letters move about when you switch keyboard layouts. Take a look a French or German for examples. I'm going to experiment with sending only scan codes (instead of virtual key codes) and see if shortcuts etc still work. This may allow me to completely ignore the OS language. Feel free to give it a go yourself by commenting out the switch statement in CharExtensions where characters are mapped to VirtualKeyCodes and try to enter shortcuts.

**Comment by JuliusSweetland** [comment]
@Ar2rZ I have updated the Russian branch with some new code which attempts to dynamically work out which physical keys to simulate when you type a character or word. Could you please pull the latest and test it out when you have a moment? It seems to work and will also deal with the OS keyboard language changing as OptiKey is running. 

I think it works correctly, but I will also be adding a new setting "Simulate physical key strokes" which will allow users to enable/disable this behaviour as modifiers (SHIFT/CTRL/ALT) + physical key simulation might confuse some users. Let me know what you think.

**Comment by Ar2rZ** [comment]
@JuliusSweetland will test this at some point during the day or then later this evening. depends on free time.
thanks.

**Comment by Ar2rZ** [comment]
did some test on win 8.1.
1. short keys working with Russian general, 
2. does not work properly with Russian mnemonic - short keys are doing completely incorrect work.
3. when is chosen English(or other language, i have Irish additionally installed) it outputs Russian chars, short keys does not work.

**Comment by JuliusSweetland** [comment]
What are Russian mnemonics?

Could you give me a step by step to reproduce issues 2&3, including which
application you are typing into, what you are typing and what is being
output, and what your OptiKey and os keyboard language are set to? Thanks.

**Comment by Ar2rZ** [comment]
Windows 8 has two Russian layouts - General and Mnemonic.
Mnemonic is when you type on english keyboard for example letter 'P' in Russian you will get russian "П". you can get a look on translit.ru to understand better.

**Comment by JuliusSweetland** [comment]
Is that a keyboard layout that can be installed? I only saw and tested
Russian.

**Comment by Ar2rZ** [comment]
so, stpes to reproduce:
- open notepad 
2. -  install Russian mnemonic language pack. 
    -  on OptiKey - win key + space so many times till it will choose  Russian Mnemonic  
    -  click some short keys. ctrl+a('ф' on Russian layout),or ctrl+c( 'c' on Russian layout), 
it will do some incorrect short keys functions. some letters will have outp put o the letter.

3. - on OptiKey _ win key + space till English or other language not Russian(probably is does not matter) is chosen
- click some short keys, it will just type in Russian on notepad.

**Comment by Ar2rZ** [comment]
auto formatting on GitHub will kill me. :) i hope you will understand my horrible writing.

**Comment by Ar2rZ** [comment]
![languages](https://cloud.githubusercontent.com/assets/14283948/10976380/6a19464c-83e3-11e5-92a8-28956f814773.png)
![russian](https://cloud.githubusercontent.com/assets/14283948/10976386/70439806-83e3-11e5-8f16-54b3610953ba.png)

**Comment by JuliusSweetland** [comment]
@Ar2rZ I have just tried the Russian Mneumonic layout and here are my observations:

I think OptiKey is pressing the correct keys, but the letters are in different positions from the standard Russian keyboard. When OptiKey is set to Russian and you select the "a" key, OptiKey simulates a key press on whichever key produces an "a", based on your current keyboard layout. This is NOT based on the position of the key shown in OptiKey. I think what you are saying is that the "a" key is in the middle row on the far left on a Russian Mneumonic keyboard, so you are selecting the OptiKey key that is in that position, which is ф. Actually you want to select the key labelled "a" which will produce the effect you expect, i.e. CTRL+"a" = select all in notepad.

![image](https://cloud.githubusercontent.com/assets/4722561/10978313/22def442-83ee-11e5-8709-7fd73fc27227.png)

I believe OptiKey is doing the right thing, but I also think we should probably avoid supporting the Russian mneumonic keyboard layout for now, as it relies on multiple keystrokes to produce some output (i.e. it looks like you can combine characters together), which is something OptiKey does NOT consider, and so the output in notepad may differ from what appears in OptiKey's scratchpad.

Let me know what you think.

**Comment by Ar2rZ** [comment]
i am agree, as mnemonic layout is used by somebody like me, who does not use Russian layout. basically lives not in Russia.
what is about English typing when OS language choosen English?

**Comment by JuliusSweetland** [comment]
If OptiKey is set to English and the OS keyboard layout is also English
then real keystrokes should be simulated for most characters. OptiKey
essentially does it's best to press keys on the OS keyboard layout. If you
can't create the character on the current OS keyboard layout then OptiKey
just send text (as Unicode characters).

**Comment by Ar2rZ** [comment]
i talking about when is OptiKey set up to Russian.
as i am understanding, this set up want type in English - OptiKey - Russain, OS -English?

**Comment by JuliusSweetland** [comment]
It shouldn't really matter which language the OS keyboard layout is set to,
OptiKey will do it's best, BUT it can only output simulated key presses for
characters which appear on the current OS keyboard layout. If you are
mainly typing Russian characters, and you think about shortcuts using
Russian letters, then the OS keyboard layout should be Russian, and OptiKey
should be set to Russian also. I will add a method of switching the OptiKey
keyboard language soon.

**Comment by Ar2rZ** [comment]
ah. OK, you find other way. then question is invalid.

**Comment by Ar2rZ** [comment]
probably we need to add English letters on the non-english layouts. some references to the short keys..

**Comment by JuliusSweetland** [comment]
Agreed. The other approach I'm thinking about is adding the ability to
switch language within OptiKey, so you have full support for Russian, then
you can switch to English for a minute to type a few words, and then switch
back to Russian, for example.

**Comment by JuliusSweetland** [comment]
@Ar2rZ Let me know once you've had a proper play with the new committed code on the Russian branch. Is there anything left to address before we can produce a pre-release with Russian for general testing?

**Comment by Ar2rZ** [comment]
i will do it, tonight. will let you know later, as had no time this week.. have to spend some time to study as well...

**Comment by JuliusSweetland** [comment]
Do it next week. I've head very little time recently also and wouldn't mind a weekend off. Shall we catch up on Monday (or later next week if you have exams coming up)?

**Comment by Ar2rZ** [comment]
exams are next month. but i have two assignment to do, need to prepare for a lab, and actually start to prepare for exams.  i have time today, made some rest from the studies.

**Comment by Ar2rZ** [comment]
question how do i perform alt+shift shirt key? (switching languages)

**Comment by JuliusSweetland** [comment]
Start with SHIFT and ALT up, then lock down ALT and select SHIFT to press
it down, then lock it down, then release it. That's one ALT+ SHIFT. Keep
cycling through SHIFT selections for the next SHIFT up/down, etc

**Comment by Ar2rZ** [comment]
on windows 7 have a problem to proper load of OptiKey. is if i choose Russian, it still showing english uk, and all menus apart the layout is self is in english.
![001](https://cloud.githubusercontent.com/assets/14283948/11007323/07d7ea68-84c1-11e5-865a-34a7ce8317bb.png)

**Comment by Ar2rZ** [comment]
![002](https://cloud.githubusercontent.com/assets/14283948/11007407/8cee933c-84c1-11e5-83ed-64c8bd9770f5.png)
![003](https://cloud.githubusercontent.com/assets/14283948/11007409/90eed582-84c1-11e5-9810-1e9f341c619a.png)

**Comment by Ar2rZ** [comment]
windows 7 - short keys does not work for general Russian at all, no output, and no short cut functions..

**Comment by Ar2rZ** [comment]
is working incorrectly on Russian mnemonic, but is same as in Win 8 incorrect..

**Comment by JuliusSweetland** [comment]
@Ar2rZ I split out the language into 2 settings; 'Keyboard language' (which controls the key layouts and dictionary to use), and 'UI language' (which controls the language used for all the Transifex strings). Is there a problem, or is this just unexpected behaviour as I forgot to mention that I'd done it?

The reason is so that you can set both languages to Russian, but then switch only the keyboard language to English to type some english words, but keep the interface in Russian. I think this was your suggestion from a couple of weeks ago.

**Comment by Ar2rZ** [comment]
sorry, so tired, that don't see anything...

**Comment by JuliusSweetland** [comment]
Let's catch up when you're not tired. I thought the Russian mnemonic
problem was something we could ignore?

**Comment by Ar2rZ** [comment]
so, looks good, but in case of use a bit tricky... as it will require assistance each time user will need to switch languages. is need to have possibility to switch output language without accessing settings.. ignore mnemonic, as just mentioned, while main keyboard short keys did not work. i actually reproduced this again. will write steps later.

**Comment by JuliusSweetland** [comment]
Yes. I'll add a new keyboard on the menu for switching languages when I get
a chance, so you can do this without assistance.

**Comment by Ar2rZ** [comment]
steps to reproduce issue with short keys.
1. open OptiKey, in my case short keys will work with Russian only,as my default is Russian.
2. open Settings
3. choose Keyboard Language other the is current(in my case it was Russian, and i switch to English), short keys want work for English
4. restart OptiKey, try Short Keys -  it will work with English only
5. open settings
6 choose your language, in my case it will be Russian, try to do short keys - does not work, but it will work with default language,with which OptiKey started.

**Comment by JuliusSweetland** [comment]
@Ar2rZ I have tried to reproduce what you describe. Here are my findings:

1. Initial setup
OS keyboard layout = Russian (standard, NOT mneumonic)
OptiKey keyboard language = Russian
Result: Typing works as expected. Shortcuts work as expected, e.g. CTRL+ф = select all

2. With OptiKey still open I switch the keyboard language to English
OS keyboard layout = Russian (standard, NOT mneumonic)
OptiKey keyboard language = English (UK)
Result: Typing works as expected. Shortcuts do NOT work as the English characters do not exist on the Russian keyboard layout. Instead OptiKey has to send these characters as plain text. This is true even when the characters look as if they are the same, e.g. the English "a" is not the same as the Russian "а" and no physical key press is generated.

3. Restarting OptiKey
OS keyboard layout = Russian (standard, NOT mneumonic)
OptiKey keyboard language = English (UK)
Result: Exactly the same result as above (in test 2). Restarting OptiKey does not change anything.

If you are seeing different behaviour after restarting OptiKey can you double check that the OS keyboard language is still Russian. If you are still seeing unexpected behaviour can you repeat the test and write out the exact steps (e.g. with Notepad focussed I select CTRL, then I select "ф" - I expect this to happen, but this happens instead"). You can also find the logs here (change the username in the path): C:\Users\**YOUR_USER_NAME**\AppData\Roaming\JuliusSweetland\OptiKey\Logs

In the logs you will see entries that describe exactly what is going on, e.g.

````
2015-11-07 09:34:22,277 [ 1] INFO  JuliusSweetland.OptiKey.Services.KeyboardOutputService: Publishing 'ф' => as virtual key code 65(0x41) (keyboard:Russian (Russia))
2015-11-07 09:30:41,639 [ 1] INFO  JuliusSweetland.OptiKey.Services.KeyboardOutputService: No virtual key code found for 'C' so publishing as text (keyboard:Russian (Russia))
````

N.B. Can you sync to get the latest from the Russian branch as I've fixed something in the logging.

**Comment by Ar2rZ** [comment]
so, is actually working as expected, 
my system is set up so, that every app is using which i chose when last used, by default is English.(i think this is a default  Windows function for multilingual setups)
even when i chose OC language Russian, after restart OtpiKey uses English as OS language .
thats why short keys stopping working,as OS language for OptiKey becomes an English.

**Comment by Ar2rZ** [comment]
i think there supposed to be way how to control OS language for OptiKey.

**Comment by JuliusSweetland** [comment]
I don't think it's possible to change the OS language from C# I'm afraid.
It would annoy people if you could.

**Comment by Ar2rZ** [comment]
the thing is if OS language as by me is an English, OptiKey will always have English by default. or you see other behavior on your system?

**Comment by JuliusSweetland** [comment]
On my system the OS keyboard language is completely unaffected by OptiKey. If I set it to English then it stays as English unless I manually change it.

Beware of keyboard shortcuts changing the language. I have disabled these as they are easy to trigger accidentally. By default I think you change language with ALT+SHIFT, which is easy to accidentally press using OptiKey as SHIFT is down by default when OptiKey starts, and you press ALT+M for the Management Console, so it's trivial to accidentally press ALT+SHIFT and change the keyboard language.

**Comment by Ar2rZ** [comment]
i will check on my system at work, as there is windows 8.1 installed, but i think it has same behavior as on windows 7.

**Comment by Ar2rZ** [comment]
try to set OS language to Russian or other language. then start OptiKey or other program/app, i think your OS language for OptiKey(other program/app) will become English.

**Comment by Ar2rZ** [comment]
where can i adjust code for unclick? as i have multiple screens, and is required to change my screens set up every time, if i want to test OptiKey properly.

**Comment by JuliusSweetland** [comment]
What do you mean by "unclick"?

**Comment by Ar2rZ** [comment]
once you click on the mouse there are two functions even three functions involved- click, hold and unclick... unless OptiKey used some other way for mouse click function.

**Comment by JuliusSweetland** [comment]
There are two keys for each mouse action; one presses down, or releases the
mouse button and so takes two selections to complete a click. This is
labeled "LEFT DOWN/UP". The other performs the whole click with one
selection and is labeled "LEFT CLICK"

**Comment by Ar2rZ** [comment]
dictionary is bind with interface layout?

**Comment by JuliusSweetland** [comment]
@Ar2rZ No, resource language

**Comment by JuliusSweetland** [comment]
@Ar2rZ Ignore my last response (deleted) - yes, dictionary is currently linked to UI language, which is the ResourceLanguage setting. This is wrong and needs to be fixed. I'll take a look tonight.

**Comment by Ar2rZ** [comment]
ok. what is still to do? in terms of make pre-release.

**Comment by JuliusSweetland** [comment]
I need to correct this small issue and then whatever else is still not
working or incomplete for Russian. Is there anything?

**Comment by Ar2rZ** [comment]
switching languages from console only.

**Comment by Ar2rZ** [comment]
but will continue to test, when will have time.

**Comment by Ar2rZ** [comment]
@JuliusSweetland  , according to mouse function(click, unclick) - 
stock on code in class MouseOutputService, method Move and left click.
position before this line  publishService.MouseMouseToPoint(point); cursor position is correct. example :{X = 907 Y = 393}) and after this line is becomes {X = -693 Y = 393} any doubt? as i can't find where is leading to..

**Comment by Ar2rZ** [comment]
ignore previous message(deleted),
i found where is problem...

**Comment by Ar2rZ** [comment]
@JuliusSweetland  
if you interested, fix for multiple screens.
* 1. For correct docking:
   * **WindowManipulationService.cs**
        * on line 1207 remove validation finalDockLeftInDp<0, as if screen attached on the left it gives  negative value for left screen width. 
* 2. For correct mouse functioning:
   * **PublishService.cs**
   * **public void MouseMouseToPoint(Point point)**
      * **change this:** 
      * var virtualScreenWidthInPixels1 = SystemParameters.VirtualScreenWidth * Graphics.DipScalingFactorX;
      * var virtualScreenHeightInPixels1 = SystemParameters.VirtualScreenHeight * Graphics.DipScalingFactorY;
      * **to this:**
      * var virtualScreenWidthInPixels = SystemParameters.WorkArea.Width * Graphics.DipScalingFactorX;
      * var virtualScreenHeightInPixels = SystemParameters.WorkArea.Height * Graphics.DipScalingFactorY;
as in second fix, **VirtualScreenWidth** gives a width of all screens together, **WorkArea.Width** given only active screen width. **VirtualScreenHeight** gives Highest  screen height, or if there will be a screen on top will give all screens height together.

**Comment by JuliusSweetland** [comment]
Thanks Arthur. I'll take a look at this when I get a chance.

**Comment by Ar2rZ** [comment]
fix for mouse function does not work for win 8.1

**Comment by JuliusSweetland** [comment]
Can you describe the problem fully in a new issue?

**Comment by Ar2rZ** [comment]
will do it later today.

**Comment by Ar2rZ** [comment]
removed all messages related to the multiple screen issue.

**Comment by JuliusSweetland** [comment]
New (2nd) pre-release of OptiKey with Russian support available here: https://github.com/OptiKey/OptiKey/releases/tag/v2.2.2_Russian_pre-release_2

I have added a new keyboard to the menu called 'Language' which allows you to change the keyboard layout and dictionary to another language. N.B. changing language does NOT change the user interface language, so if you have OptiKey set up with UILanguage=English and KeyboardLanguage=English, then change to Russian, for example, your keyboard layout and dictionary will change to Russian, but your user interface strings will stay in English.

The other major change is the already discussed (above) changes to the way physical keys are simulated. In short OptiKey will only attempt to simulate a physical key press if the character you select in OptiKey can be made from a key combination on your Windows keyboard layout.

Let me know what you think.

**Comment by Ar2rZ** [comment]
thanks, Julius.
will look at this tonight.

**Comment by Ar2rZ** [comment]
Hi Julius, i try on laptop with Windows 8.1 - installed pre-release version 2.2.2, is very very slow... checked on task manager performance Processor is just stable 18-20%, RAM just 46%, disc 0%. not able to work optikey, is just to slow.

**Comment by Ar2rZ** [comment]
try on my machine, is looks ok. but i prefer to have written language function on main keyboard(i think is not critical at the moment, as we have one button on main keyboard unoccupied with function), another thing - once more languages will be introduced, it will be no space(and no need for user)to be displayed all languages.

**Comment by JuliusSweetland** [comment]
>Hi Julius, i try on laptop with Windows 8.1 - installed pre-release version 2.2.2, is very very slow... checked on task manager performance Processor is just stable 18-20%, RAM just 46%, disc 0%. not able to work with optikey, is just to slow.

Hi Arthur - can you try again with;
a) The new version (2.2.2 pre-release), but in the management screen go to the Other tab and turn OFF "Simulate physical keystrokes when typing characters"
b) The previous version

Can you let me know if either improves the performance of OptiKey.

> try on my machine, is looks ok. but i prefer to have written language function on main keyboard(i think is not critical at the moment, as we have one button on main keyboard unoccupied with function), another thing - once more languages will be introduced, it will be no space(and no need for user)to be displayed all languages.

I need to think about this further. For now I'd like to get to a point where the performance is ok and you think all functionality is there. We can then consider ways to improve the key layout and how quickly you can switch between languages. The language keyboard is just one idea.

**Comment by JuliusSweetland** [comment]
I'm keen to get Russian merged back into master and released with a few other new features and bug fixes this coming week. Could someone please let me know ff the Russian pre-release is useable in light of Arthur's performance feedback? Does anyone else have problems?

**Comment by Ar2rZ** [comment]
Hi Julius, sorry for delays. i had no really time to test on laptop, as have limited access to this laptop.
is looks better with new release.

**Comment by JuliusSweetland** [comment]
No problem. So you're happy for me to release in its current state (pre
release 2.2.2 plus a couple of small bug fixes)?

**Comment by Ar2rZ** [comment]
Yop. i am happy with this release. thanks. :)

**Comment by JuliusSweetland** [comment]
Great. I'll release tonight.

**Comment by JuliusSweetland** [comment]
Actually, I still need a couple of translations done. @Ar2rZ could you please translate the following 2 strings into Russian:

1."Suppress mouse magnification for scrolling actions"
2."Conversation keyboard thickness (left,top,right,bottom in px)"

**Comment by JuliusSweetland** [comment]
3rd pre-release available for testing here: https://github.com/OptiKey/OptiKey/releases/tag/v2.2.2_Russian_pre-release_3

FYI This is hopefully the last release candidate. I just have to include the above 2 translations in Russian and French and then I'll release this.

**Comment by Ar2rZ** [comment]
i will translate later today.

**Comment by Ar2rZ** [comment]
Additional strings are  translated.

**Comment by JuliusSweetland** [comment]
Thanks Arthur. I'll pull and release later on this evening. Beer permitting.

**Comment by JuliusSweetland** [comment]
Hi Arthur - where are the translated strings? I cannot see them.

**Comment by Ar2rZ** [comment]
hi Julius,

can you check, whether my pool request is correct.
as seems to be there are some other unexpected changes included...

**Comment by JuliusSweetland** [comment]
I'll take a look later, thank you.

**Comment by JuliusSweetland** [comment]
Final version available here: https://github.com/OptiKey/OptiKey/releases/tag/v2.2.2

I will release this as a full release tomorrow unless anyone finds any problems with it. Closing this issue as Russian should hopefully be completed. Great work everyone and thank you.
